src/blocksparsearrayinterface/map.jl

```diff
@@ -94,23 +96,17 @@ function map_zero_dim! end
   return a_dest
 end
 
-# TODO: Decide what to do with these.
+# TODO: Do we need this function or can we just use `map`?
+# Probably it should be a special version of `map` where we
+# specify the function preserves zeros, i.e.
+# `map(f, a; preserves_zeros=true)` or `@preserves_zeros map(f, a)`.
 function map_stored_blocks(f, a::AbstractArray)
-  # TODO: Implement this as:
-  # ```julia
-  # mapped_blocks = SparseArraysInterface.map_stored(f, blocks(a))
-  # BlockSparseArray(mapped_blocks, axes(a))
-  # ```
-  # TODO: `block_stored_indices` should output `Indices` storing
-  # the stored Blocks, not a `Dictionary` from cartesian indices
-  # to Blocks.
-  bs = collect(eachblockstoredindex(a))
-  ds = map(b -> f(@view(a[b])), bs)
-  # We manually specify the block type using `Base.promote_op`
-  # since `a[b]` may not be inferrable. For example, if `blocktype(a)`
-  # is `Diagonal{Float64,Vector{Float64}}`, the non-stored blocks are `Matrix{Float64}`
-  # since they can't necessarily by `Diagonal` if there are rectangular blocks.
-  mapped_blocks = Dictionary{eltype(bs),eltype(ds)}(bs, ds)
-  # TODO: Use `similartype(typeof(a), eltype(eltype(mapped_blocks)))(...)`.
-  return BlockSparseArray(mapped_blocks, axes(a))
+  blocks_a = blocks(a)
+  stored_indices = collect(eachstoredindex(blocks_a))
+  stored_blocks = map(I -> f(blocks_a[I]), stored_indices)
+  blocks_a′ = SparseArrayDOK{eltype(stored_blocks)}(undef_blocks, axes(a))
```

Comment:
Can we immediately construct a `BlockSparseArray` instead?

Reply:
That's how I had written it, partially this rewrite was to try out this new `SparseArrayDOK` constructor, but probably writing it in terms of `BlockSparseArray` is better, I'll change it back.